beamer/config/commands.py

```diff
@@ -0,0 +1,133 @@
+
+        case LpAdded():
+            if event.event_address == deployment.chain.contracts["RequestManager"].address:
+                config.request_manager.whitelist.add(event.lp)
+            elif event.event_address == deployment.chain.contracts["FillManager"].address:
+                config.fill_manager.whitelist.add(event.lp)
+            else:
+                raise ValueError(f"event from an unexpected address: {event}")
+
+        case LpRemoved():
+            if event.event_address == deployment.chain.contracts["RequestManager"].address:
+                config.request_manager.whitelist.remove(event.lp)
+            elif event.event_address == deployment.chain.contracts["FillManager"].address:
+                config.fill_manager.whitelist.remove(event.lp)
+            else:
+                raise ValueError(f"event from an unexpected address: {event}")
+
+
+@click.group()
+def config() -> None:
+    pass
+
+
+@config.command("read")
+@click.option(
+    "--rpc-file",
+    type=click.Path(exists=True, file_okay=True, dir_okay=False, path_type=Path),
+    required=True,
+    help="Path to the RPC config file.",
+)
+@click.option(
+    "--abi-dir",
+    type=click.Path(exists=True, file_okay=False, dir_okay=True, path_type=Path),
+    required=True,
+    help="Path to the directory with contract ABIs.",
+)
+@click.option(
+    "--artifact",
+    type=click.Path(exists=True, file_okay=True, dir_okay=False, path_type=Path),
+    required=True,
+    help="Path to the deployment artifact.",
+)
+@click.argument("state_path", type=click.Path(file_okay=True, dir_okay=False, path_type=Path))
+def read(
+    rpc_file: Path, abi_dir: Path, artifact: Path, state_path: Path
+) -> None:  # pylint: disable=unused-argument
+    """Read latest contract configuration state from the chain and store it into STATE_PATH."""
+    beamer.util.setup_logging(log_level="DEBUG", log_json=False)
+
+    rpc_info = beamer.deploy.config.load_rpc_info(rpc_file)
+    deployment = Deployment.from_file(artifact)
+
+    assert deployment.chain is not None
+    chain_id = deployment.chain.chain_id
+    url = rpc_info[chain_id]
+    w3 = make_web3(url)
+    assert w3.eth.chain_id == chain_id
+    log.info("Connected to RPC", url=url)
+
+    request_manager = deployment.obtain_contract(w3, "chain", "RequestManager")
+    fill_manager = deployment.obtain_contract(w3, "chain", "FillManager")
+    # Here we assume that the configuration has not changed between the two
+    # deployment blocks.
+    start_block = max(
+        deployment.chain.contracts["RequestManager"].deployment_block,
+        deployment.chain.contracts["FillManager"].deployment_block,
+    )
+    fetcher = EventFetcher(
+        w3, (request_manager, fill_manager), start_block=start_block, confirmation_blocks=0
+    )
+
+    if state_path.exists():
+        config = Configuration.from_file(state_path)
+    else:
+        config = Configuration.initial(chain_id, start_block)
+
+    events = fetcher.fetch()
```

Comment:
With this call we cannot be really sure that the `synced_block` is actually latest. I know this is a bit tricky, as while replaying the events the blockchain progresses. 
The fetch() can exit without progressing to the latest block.

The code is good as is, but it can happen that `fetch()` exited early. Then the synced block can be actually far away from the actual latest state. 

One scenario could be that at the very top a `RequestException` is raised upon fetching the latest block data. 

The output would be an empty state file. This is noted with the block number. However the block number doesn't really tell me much when seeing it. Even further, if we were to automize things then it could become quite critical. 

I don't have a "good" solution for now, but maybe adding the timestamp of the block in a readable format could help humans at least.

Regarding automization, it actually has to be handled in the `write` command that the state is not up to date.

Reply:
> Regarding automization, I assume it should actually be handled by the `write` command, noticing that the state is not uptodate and then fetch the latest state. However, there the write command need to be more time sensitive about the synced block then.

That is precisely my thinking. The read side can only do so much, but given that there will always be some time between you read from chain and write to chain, it is reasonable to move the responsibility to the write side. We have to check anyway if there were configuration changes since the last state file was written. If there were, the write command should just exit with an error and advise you to run the read command again. In a way, we don't particularly care if there are 200 blocks between the block from the state file and the current latest block, if there were no state changes in between.

Theoretically, it could happen that the block from the file is too far in the past so we should not waste time fetching events up to the latest block, only to exit with an error. What we could do is check whether the last synced block is older than e.g. 1 day and if yes, just exit with an error and require a new sync via the read command.